data/tutorials/language/0lg_09_env.md

```diff
@@ -0,0 +1,610 @@
+
+However, all future expressions will use the new value of `d` (`10`), as shown here:
+
+```
+# let g = d * 3;;
+val g : int = 30
+```
+
+Partially applying parameters to a function creates a new closure. The environment is updated, but the function is unchanged.
+```ocaml
+# let max_42 = max 42;;
+val max_42 : int -> int = <fun>
+```
+
+Inside the `max_42` function, the environment contains an additional binding between the first parameter of `max` and the value 42.
+
+## Recursive Functions
+
+In order to perform iterated computations, a function may call itself. Such a function is called _recursive_. In OCaml, recursive functions must be defined and explicitly declared by using `let rec`. It is not possible to accidentally create recursion loops between functions. As a consequence, recursive functions can't be anonymous.
+
+The classic (and very inefficient) way to present recursion is using the function that computes the [Fibonacci Sequence](https://en.wikipedia.org/wiki/Fibonacci_sequence).
+```ocaml
+# let rec fibo n = if n <= 1 then n else fibo (n - 1) + fibo (n - 2);;
+val fibo : int -> int = <fun>
+
+# List.init 10 Fun.id |> List.map fibo;;
+- : int list = [0; 1; 1; 2; 3; 5; 8; 13; 21; 34]
+```
+
+This version of `fibo` is inefficient because the number of recursive calls created doubles at each call, which creates exponential growth.
+
+This version does a better job:
+```ocaml
+# let rec fibo m n i = if i = 0 then m else fibo n (n + m) (i - 1);;
+val fibo : int -> int -> int -> int = <fun>
+
+# let fibo = fibo 0 1;;
+val fibo : int -> int = <fun>
+
+# List.init 10 Fun.id |> List.map fibo;;
+- : int list = [0; 1; 1; 2; 3; 5; 8; 13; 21; 34]
+```
+
+The first version takes two extra parameters: the two previously computed Fibonacci numbers.
+
+The second version uses the two first Fibonacci numbers as initial values. There is nothing to be computed when returning from a recursive call, so this enables the compiler to perform an optimisation called “tail call elimination.” This turns recursivity into imperative iteration in the generated native code and leads to much-improved performances.
+
+## Multiple Arguments Functions
+
+### Arrow Types and Mutiple Arguments
+
+Until now, functions with several arguments had a type looking like this: `a₁ -> a₂ -> a₃ -> ⋯ -> b`
+
+Where `a₁` is the type of the first argument, `a₂` is the type of the second argument, etc., and `b` is the type of the result. However, this type isn't displayed as it is. In reality, this type is the following: `a₁ -> (a₂ -> (a₃ -> ⋯ -> b))`. The arrow symbol is a [binary operator](https://en.wikipedia.org/wiki/Binary_operation). Here is how this can be observed:
```

Comment:
```suggestion
Where `a₁` is the type of the first argument, `a₂` is the type of the second argument, etc., and `b` is the type of the result. However, this type isn't displayed as it is. In reality, this type is the following: `a₁ -> (a₂ -> (a₃ -> ⋯ -> b))`. The arrow symbol is a [binary operator](https://en.wikipedia.org/wiki/Binary_operation). Here is how this can be written:
```
Should this stay "observed" or changed to "written"?